common/cryptfs.c

```diff
@@ -546,47 +619,115 @@ cryptfs_write_zeros(char *crypto_blkdev, size_t size)
 	return ret;
 }
 
-static char *
-cryptfs_setup_volume_integrity_new(const char *label, const char *real_blkdev,
-				   const char *meta_blkdev, const char *key, unsigned long fs_size)
+char *
+cryptfs_setup_volume_new(const char *label, const char *real_blkdev, const char *key,
+			 const char *meta_blkdev, cryptfs_mode_t mode)
 {
+	int fd;
+	// The file system size in sectors
+	uint64_t fs_size;
+
+	char *crypto_blkdev = NULL, *integrity_blkdev = NULL;
+	char *crypto_key = NULL, *integrity_key = NULL;
+	size_t crypto_key_len = 0, integrity_key_len = 0;
+	char *integrity_dev_label = NULL;
 	bool initial_format = false;
-	char *crypto_blkdev = NULL, *integrity_dev = NULL;
-	char *integrity_dev_label = mem_printf("%s-%s", label, "integrity");
-	TRACE("cryptfs_setup_volume_integrity_new");
+	bool encrypt = true, integrity = false, stacked = false;
+
+	/* do parameter validation */
+	IF_NULL_RETVAL_ERROR(label, NULL);
+	IF_NULL_RETVAL_ERROR(real_blkdev, NULL);
+	IF_NULL_RETVAL_ERROR(key, NULL);
+
+	switch (mode) {
+	case CRYPTFS_MODE_AUTHENC:
+		IF_NULL_RETVAL_ERROR(meta_blkdev, NULL);
+		crypto_key_len = AUTHENC_HEXKEY_LEN;
+		integrity_key_len = 0;
+		IF_TRUE_RETVAL_ERROR(strlen(key) != AUTHENC_HEXKEY_LEN, NULL);
+		stacked = true;
+		integrity = true;
+		break;
+	case CRYPTFS_MODE_ENCRYPT_ONLY:
+		IF_TRUE_RETVAL(strlen(key) != CRYPTO_HEXKEY_LEN, NULL);
+		crypto_key_len = CRYPTO_HEXKEY_LEN;
+		integrity_key_len = 0;
+		break;
+	case CRYPTFS_MODE_INTEGRITY_ENCRYPT:
+		IF_TRUE_RETVAL(strlen(key) != CRYPTO_HEXKEY_LEN + INTEGRITY_HEXKEY_LEN, NULL);
+		crypto_key_len = CRYPTO_HEXKEY_LEN;
+		integrity_key_len = INTEGRITY_HEXKEY_LEN;
+		integrity = true;
+		break;
+	case CRYPTFS_MODE_INTEGRITY_ONLY:
+		IF_TRUE_RETVAL(strlen(key) != INTEGRITY_HEXKEY_LEN, NULL);
+		crypto_key_len = 0;
+		integrity_key_len = INTEGRITY_HEXKEY_LEN;
+		integrity = true;
+		encrypt = false;
+		break;
+	default:
+		ERROR("Unsupported type and or parameter combination");
+		return NULL;
+	}
+	/* validated parameters */
 
-	/* check if meta device is initialized */
-	initial_format = get_provided_data_sectors(meta_blkdev) != fs_size;
+	/*
+	 * Use the first 128 hex digits (64 byte) of master key for 512 bit xts mode
+	 * or full master key for AUTHENC mode
+	 */
+	crypto_key = crypto_key_len > 0 ? mem_strndup(key, crypto_key_len) : NULL;
 
-	if (create_integrity_blk_dev(real_blkdev, meta_blkdev, integrity_dev_label, fs_size) < 0) {
-		DEBUG("create_integrity_blk_dev failed!");
-		mem_free0(integrity_dev_label);
+	/* Use the following 64 hex digits (32 byte) of master key for 256 bit hmac */
+	integrity_key =
```

Comment:
IÄd suggest adding "if(integrity_key_len > 0)" here to avoid call to strdup with zero length?